modules/tts/tts-utils/tts-adapter-google/src/main/java/org/daisy/pipeline/tts/google/impl/GoogleRestTTSEngine.java

```diff
@@ -0,0 +1,226 @@
+		mAudioFormat = audioFormat;
+		mRequestScheduler = new ExponentialBackoffScheduler<Schedulable>();
+		mRequestBuilder = new GoogleRequestBuilder(apiKey);
+		
+	}
+
+	@Override
+	public Collection<AudioBuffer> synthesize(String sentence, XdmNode xmlSentence,
+			Voice voice, TTSResource threadResources, AudioBufferAllocator bufferAllocator, boolean retry)
+					throws SynthesisException,InterruptedException, MemoryException {
+		
+		if (sentence.length() > 5000) {
+			throw new SynthesisException("The number of characters in the sentence must not exceed 5000.");
+		}
+
+		Collection<AudioBuffer> result = new ArrayList<AudioBuffer>();
+
+		// the sentence must be in an appropriate format to be inserted in the json query
+		// it is necessary to wrap the sentence in quotes and add backslash in front of the existing quotes
+		
+		String adaptedSentence = "";
+		
+		for (int i = 0; i < sentence.length(); i++) {
+			if (sentence.charAt(i) == '"') {
+				adaptedSentence = adaptedSentence + '\\' + sentence.charAt(i);
+			}
+			else {
+				adaptedSentence = adaptedSentence + sentence.charAt(i);
+			}
+		}
+		
+		String languageCode;
+		String name;
+		int indexOfSecondHyphen;
+
+		if (voice != null) {
+			//recovery of the language code in the name of the voice
+			indexOfSecondHyphen = voice.name.indexOf('-', voice.name.indexOf('-') + 1);
+			languageCode = voice.name.substring(0, indexOfSecondHyphen);
+			name = voice.name;
+		}
+		else {
+			// by default the voice is set to English
+			languageCode = "en-GB";
+			name = "en-GB-Standard-A";
+		}
+		
+		
+		
+		
+		try {
+			Request speechRequest = mRequestBuilder.newRequest()
+					.withSampleRate((int)mAudioFormat.getSampleRate())
+					.withAction(GoogleRestAction.SPEECH)
+					.withLanguageCode(languageCode)
+					.withVoice(name)
+					.withText(adaptedSentence)
+					.build();
+			UUID requestUuid = mRequestScheduler.add(()->{
+				try {
+					BufferedReader br = new BufferedReader(new InputStreamReader(speechRequest.send(), "utf-8"));
+					StringBuilder response = new StringBuilder();
+					String inputLine;
+					while ((inputLine = br.readLine()) != null) {
+						response.append(inputLine.trim());
+					}
+					br.close();
+
+					// the answer is encoded in base 64, so it must be decoded
+					byte[] decodedBytes = Base64.getDecoder().decode(response.toString().substring(18, response.length()-2));
+
+					AudioBuffer b = bufferAllocator.allocateBuffer(decodedBytes.length);
+					b.data = decodedBytes;
+					result.add(b);
+				} catch (IOException | MemoryException e) {
+					try {
+						if (speechRequest.getConnection().getResponseCode() == 429) {
+							throw new RecoverableError("Exceeded quotas", e);
+						} else {
+							throw new FatalError(e);
+						}
+					} catch (IOException responseCodeError) {
+						throw new FatalError("could not retrieve response code for request", responseCodeError);
+					}
+				}
+			});
+			
+			mRequestScheduler.launch(requestUuid);
```

Comment:
Is there any reason why you would not "launch" the request immediately after adding it? If "add" and "launch" are combined there is no need for the UUID.

Reply:
I keep a fix storage process for all scheduler (with the add function implemented in the abstract scheduler class), and let futur implementations of the scheduler class decide what to do when launching a specific request with that approach (or use existing scheduler.
if you think its a problem i can change that

Reply:
Not sure I get it.